app/src/behaviors/behavior_macro.c

```diff
@@ -0,0 +1,187 @@
+};
+
+struct behavior_macro_trigger_state {
+    uint32_t wait_ms;
+    uint32_t tap_ms;
+    enum behavior_macro_mode mode;
+    uint8_t start_index;
+    uint8_t count;
+};
+
+struct behavior_macro_state {
+    struct behavior_macro_trigger_state release_state;
+
+    uint8_t press_bindings_count;
+};
+
+struct behavior_macro_config {
+    uint32_t default_wait_ms;
+    uint32_t default_tap_ms;
+    uint8_t count;
+    struct zmk_behavior_binding bindings[];
+};
+
+#define TAP_MODE DT_LABEL(DT_INST(0, zmk_macro_control_mode_tap))
+#define PRESS_MODE DT_LABEL(DT_INST(0, zmk_macro_control_mode_press))
+#define REL_MODE DT_LABEL(DT_INST(0, zmk_macro_control_mode_release))
+
+#define TAP_TIME DT_LABEL(DT_INST(0, zmk_macro_control_tap_time))
+#define WAIT_TIME DT_LABEL(DT_INST(0, zmk_macro_control_wait_time))
+#define WAIT_REL DT_LABEL(DT_INST(0, zmk_macro_pause_for_release))
+
+#define ZM_IS_NODE_MATCH(a, b) (strcmp(a, b) == 0)
+#define IS_TAP_MODE(dev) ZM_IS_NODE_MATCH(dev, TAP_MODE)
+#define IS_PRESS_MODE(dev) ZM_IS_NODE_MATCH(dev, PRESS_MODE)
+#define IS_RELEASE_MODE(dev) ZM_IS_NODE_MATCH(dev, REL_MODE)
+
+#define IS_TAP_TIME(dev) ZM_IS_NODE_MATCH(dev, TAP_TIME)
+#define IS_WAIT_TIME(dev) ZM_IS_NODE_MATCH(dev, WAIT_TIME)
+#define IS_PAUSE(dev) ZM_IS_NODE_MATCH(dev, WAIT_REL)
+
+static bool handle_control_binding(struct behavior_macro_trigger_state *state,
+                                   const struct zmk_behavior_binding *binding) {
+    if (IS_TAP_MODE(binding->behavior_dev)) {
+        state->mode = MACRO_MODE_TAP;
+        LOG_DBG("macro mode set: tap");
+    } else if (IS_PRESS_MODE(binding->behavior_dev)) {
+        state->mode = MACRO_MODE_PRESS;
+        LOG_DBG("macro mode set: press");
+    } else if (IS_RELEASE_MODE(binding->behavior_dev)) {
+        state->mode = MACRO_MODE_RELEASE;
+        LOG_DBG("macro mode set: release");
+    } else if (IS_TAP_TIME(binding->behavior_dev)) {
+        state->tap_ms = binding->param1;
+        LOG_DBG("macro tap time set: %d", state->tap_ms);
+    } else if (IS_WAIT_TIME(binding->behavior_dev)) {
+        state->wait_ms = binding->param1;
+        LOG_DBG("macro wait time set: %d", state->wait_ms);
+    } else {
+        return false;
+    }
+
+    return true;
+}
+
+static int behavior_macro_init(const struct device *dev) {
+    const struct behavior_macro_config *cfg = dev->config;
+    struct behavior_macro_state *state = dev->data;
+    state->press_bindings_count = cfg->count;
+    state->release_state.start_index = cfg->count;
+    state->release_state.count = 0;
+
+    LOG_DBG("Precalculate initial release state:");
+    for (int i = 0; i < cfg->count; i++) {
+        if (handle_control_binding(&state->release_state, &cfg->bindings[i])) {
+            // Updated state used for initial state on release.
+        } else if (IS_PAUSE(cfg->bindings[i].behavior_dev)) {
+            state->release_state.start_index = i + 1;
+            state->release_state.count = cfg->count - i - 1;
```

Comment:
maybe simplification? IMO this is more obvious than having a `- 1` in the expression.
```suggestion
            state->release_state.count = cfg->count - state->release_state.start_index;
```